### PR TITLE
Saves receiver guid when creating transfer agreement

### PR DESCRIPTION
**Pull request by tnickelsen**
Need integration tests. Current tests fail because we are trying to create a DepositEndpoint in the wallet with the same public key as the one generated under CreateWalletDepositEndpoint. It violates the unique constraint in the wallet.

### CODE REVIEW

File: domains/transfer-agreements/TransferAgreements.API/API.IntegrationTests/Testcontainers/WalletContainer.cs [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+        walletContainer = new Lazy<IContainer>(() =>
+        {
+            var postgresConnectionString = $"Host={postgresContainer.IpAddress};Port=5432;Database=postgres;Username=postgres;Password=postgres";
+
+            return new ContainerBuilder()
+                .WithImage("ghcr.io/project-origin/wallet-server:0.1.3")
+                .WithPortBinding(80, true)
+                .WithCommand("--serve", "--migrate")
+                .WithEnvironment("ConnectionStrings__Database", postgresConnectionString)
+                .WithEnvironment("ServiceOptions__EndpointAddress", $"http://postgres:5432/")
```
Comment (duizer) [inline domains/transfer-agreements/TransferAgreements.API/API.IntegrationTests/Testcontainers/WalletContainer.cs:38]: This is probably not an issue wrt. the current usage of the Wallet by the TransferAgreement domain, but this is not the correct setting for ServiceOptions:EndpointAddress. It needs to be the endpoint reachable on the host, which requires that you know the host port assigned to the container. For this reason we have hardcoded the host port in the certificates domain to 7890, which means that the endpoint address becomes "http://localhost:7890".

Comment (duizer) [reply domains/transfer-agreements/TransferAgreements.API/API.IntegrationTests/Testcontainers/WalletContainer.cs:38]: See https://github.com/Energinet-DataHub/energy-origin/blob/main/domains/certificates/Query.API/API.IntegrationTests/Testcontainers/ProjectOriginStack.cs

File: domains/transfer-agreements/TransferAgreements.API/API.IntegrationTests/Controllers/TransferAgreementsControllerTest.cs [hunk 2]
```diff
@@ -64,14 +67,16 @@ await factory.SeedData(new List<TransferAgreement>()
                 SenderId = senderId,
                 SenderName = "nrgi A/S",
                 SenderTin = "44332211",
-                ReceiverTin = "12345678"
+                ReceiverTin = "12345678",
+                ReceiverReference = Guid.Parse("5b24891b-1ab7-48c0-b747-9341a7dc6eec")
```
Comment (ckr123) [inline domains/transfer-agreements/TransferAgreements.API/API.IntegrationTests/Controllers/TransferAgreementsControllerTest.cs:71]: Why is this hardcoded? Could it just be `Guid.NewGuid()`?

If it not required to be hardcoded, then i would suggest changing it everywhere to `NewGuid`

File: domains/transfer-agreements/TransferAgreements.API/API.IntegrationTests/Factories/TransferAgreementsApiWebApplicationFactory.cs [hunk 3]
```diff
@@ -30,19 +30,29 @@ public class TransferAgreementsApiWebApplicationFactory : WebApplicationFactory<
 
     Task IAsyncLifetime.DisposeAsync() => testContainer.DisposeAsync().AsTask();
 
+    public string? WalletUrl { get; set; }
```
Comment (ckr123) [inline domains/transfer-agreements/TransferAgreements.API/API.IntegrationTests/Factories/TransferAgreementsApiWebApplicationFactory.cs:33]: Shouldnt there always be a wallet url?

File: domains/transfer-agreements/TransferAgreements.API/API.IntegrationTests/Testcontainers/WalletContainer.cs [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+public class WalletContainer : IAsyncLifetime
+{
+    private readonly Lazy<IContainer> walletContainer;
+    private readonly PostgreSqlContainer postgresContainer;
+
+    public WalletContainer()
+    {
+        postgresContainer = new PostgreSqlBuilder()
+            .WithDatabase("postgres")
```
Comment (ckr123) [inline domains/transfer-agreements/TransferAgreements.API/API.IntegrationTests/Testcontainers/WalletContainer.cs:18]: Isnt it set by default these values?

File: domains/transfer-agreements/TransferAgreements.API/API/Services/WalletDepositEndpointService.cs [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+using Microsoft.Extensions.Options;
+using ProjectOrigin.WalletSystem.V1;
+
+namespace API.Services;
+public class WalletDepositEndpointService : IWalletDepositEndpointService
+{
+    private readonly IOptions<ProjectOriginOptions> projectOriginOptions;
+    private readonly ILogger<WalletDepositEndpointService> logger;
+
+    public WalletDepositEndpointService(IOptions<ProjectOriginOptions> grOptions, ILogger<WalletDepositEndpointService> logger)
```
Comment (ckr123) [inline domains/transfer-agreements/TransferAgreements.API/API/Services/WalletDepositEndpointService.cs:18]: What does the `gr`stands for? Can we be more specific in the naming?

File: domains/transfer-agreements/TransferAgreements.API/API.IntegrationTests/Controllers/TransferAgreementsControllerTest.cs [hunk 6]
```diff
@@ -18,16 +19,18 @@
 namespace API.IntegrationTests.Controllers;
 
 [UsesVerify]
-public class TransferAgreementsControllerTests : IClassFixture<TransferAgreementsApiWebApplicationFactory>
+public class TransferAgreementsControllerTests : IClassFixture<TransferAgreementsApiWebApplicationFactory>, IClassFixture<WalletContainer>
```
Comment (ckr123) [inline domains/transfer-agreements/TransferAgreements.API/API.IntegrationTests/Controllers/TransferAgreementsControllerTest.cs:22]: Would be nice with some test to cover the Acceptance criterias

File: domains/transfer-agreements/TransferAgreements.API/API.IntegrationTests/Controllers/TransferAgreementHistoryEntriesControllerTests.cs [hunk 7]
```diff
@@ -1,40 +1,59 @@
...
+using Xunit.Abstractions;
 
 namespace API.IntegrationTests.Controllers;
 
 [UsesVerify]
-public class TransferAgreementHistoryEntriesControllerTests : IClassFixture<TransferAgreementsApiWebApplicationFactory>
+public class TransferAgreementHistoryEntriesControllerTests : IClassFixture<TransferAgreementsApiWebApplicationFactory>, IClassFixture<WalletContainer>
 {
     private readonly TransferAgreementsApiWebApplicationFactory factory;
+    private readonly ITestOutputHelper testOutputHelper;
```
[the remark this answers is not in the document]
Comment (sahma19) [reply domains/transfer-agreements/TransferAgreements.API/API.IntegrationTests/Controllers/TransferAgreementHistoryEntriesControllerTests.cs:23]: I used it in order to console output test results, in the pipeline, for debugging purposes. i just forgot to remove it again. It is removed now.

File: domains/transfer-agreements/TransferAgreements.API/API.IntegrationTests/Controllers/TransferAgreementHistoryEntriesControllerTests.cs [hunk 8]
```diff
@@ -131,10 +152,18 @@ public async Task GetHistoryEntriesForTransferAgreement_ShouldNotContainActorNam
...
             new DateTimeOffset(2123, 3, 3, 3, 3, 3, TimeSpan.Zero).ToUnixTimeSeconds(),
             new DateTimeOffset(2124, 4, 4, 4, 4, 4, TimeSpan.Zero).ToUnixTimeSeconds(),
-            "12345678"
+            "12345678",
+            Some.Base64EncodedWalletDepositEndpoint
         );
 
         var createRequest = await senderClient.PostAsJsonAsync("api/transfer-agreements", transferAgreement);
+
+        if (!createRequest.IsSuccessStatusCode)
```
[the remark this answers is not in the document]
Comment (sahma19) [reply domains/transfer-agreements/TransferAgreements.API/API.IntegrationTests/Controllers/TransferAgreementHistoryEntriesControllerTests.cs:161]: I used it in order to console output test results, in the pipeline, for debugging purposes. i just forgot to remove it again. It is removed now.

File: domains/transfer-agreements/TransferAgreements.API/API.IntegrationTests/Controllers/TransferAgreementHistoryEntriesControllerTests.cs [hunk 9]
```diff
@@ -1,40 +1,59 @@
...
     {
         var senderId = Guid.NewGuid();
-        var authenticatedClient = factory.CreateAuthenticatedClient(sub: senderId.ToString());
+
+        var result = await factory
+            .CreateAuthenticatedClient(sub: senderId.ToString())
+            .PostAsync("api/transfer-agreements/wallet-deposit-endpoint", null);
+
+        var resultData = JsonConvert.DeserializeObject<Dictionary<string, string>>(await result.Content.ReadAsStringAsync());
+        var base64String = resultData?["result"];
```
[the remark this answers is not in the document]
Comment (sahma19) [reply domains/transfer-agreements/TransferAgreements.API/API.IntegrationTests/Controllers/TransferAgreementHistoryEntriesControllerTests.cs:43]: Changes

File: domains/transfer-agreements/TransferAgreements.API/API.IntegrationTests/Controllers/TransferAgreementHistoryEntriesControllerTests.cs [hunk 10]
```diff
@@ -1,40 +1,59 @@
...
         var transferAgreement = new CreateTransferAgreement(
             new DateTimeOffset(2123, 3, 3, 3, 3, 3, TimeSpan.Zero).ToUnixTimeSeconds(),
             new DateTimeOffset(2124, 4, 4, 4, 4, 4, TimeSpan.Zero).ToUnixTimeSeconds(),
-            "12345678"
+            "12345456",
+            base64String
         );
 
-        var createRequest = await authenticatedClient.PostAsJsonAsync("api/transfer-agreements", transferAgreement);
+        var createRequest = await factory.CreateAuthenticatedClient(sub: senderId.ToString()).PostAsJsonAsync("api/transfer-agreements", transferAgreement);
```
[the remark this answers is not in the document]
Comment (sahma19) [reply domains/transfer-agreements/TransferAgreements.API/API.IntegrationTests/Controllers/TransferAgreementHistoryEntriesControllerTests.cs:52]: A receiver needs to create the wallet deposit endpoint first. then a sender needs to use that wallet deposit endpoint to create the transfer agreement between him and the sender. that is why they need their own client. I adjusted the tests variables, to have names that are easier to understand,

File: domains/transfer-agreements/TransferAgreements.API/API.IntegrationTests/Controllers/TransferAgreementHistoryEntriesControllerTests.cs [hunk 11]
```diff
@@ -1,40 +1,58 @@
...
 
     [Fact]
     public async Task Create_ShouldGenerateHistoryEntry_WhenTransferAgreementIsCreated()
     {
         var senderId = Guid.NewGuid();
-        var authenticatedClient = factory.CreateAuthenticatedClient(sub: senderId.ToString());
+
+        var result = await factory
+            .CreateAuthenticatedClient(sub: senderId.ToString())
+            .PostAsync("api/transfer-agreements/wallet-deposit-endpoint", null);
```
[the remark this answers is not in the document]
Comment (sahma19) [reply domains/transfer-agreements/TransferAgreements.API/API.IntegrationTests/Controllers/TransferAgreementHistoryEntriesControllerTests.cs:40]: The post request contains no body. so it can't be done like that i think. but i have changed the deserialization to be more neat.

File: domains/transfer-agreements/TransferAgreements.API/API.IntegrationTests/Testcontainers/WalletContainer.cs (hunk 4 above)
Comment (sahma19) [reply domains/transfer-agreements/TransferAgreements.API/API.IntegrationTests/Testcontainers/WalletContainer.cs:18]: You are right. I have removed the specification of the default values.

File: domains/transfer-agreements/TransferAgreements.API/API.IntegrationTests/Controllers/TransferAgreementsControllerTest.cs (hunk 6 above)
Comment (sahma19) [reply domains/transfer-agreements/TransferAgreements.API/API.IntegrationTests/Controllers/TransferAgreementsControllerTest.cs]: I have added a separate test for checking the database for the receiverreferenceguid not being empty

File: domains/transfer-agreements/TransferAgreements.API/API.IntegrationTests/Controllers/TransferAgreementsControllerTest.cs (hunk 2 above)
Comment (sahma19) [reply domains/transfer-agreements/TransferAgreements.API/API.IntegrationTests/Controllers/TransferAgreementsControllerTest.cs]: Changed to use Guid.NewGuid()

File: domains/transfer-agreements/TransferAgreements.API/API.IntegrationTests/Controllers/TransferAgreementHistoryEntriesControllerTests.cs (hunk 11 above)
[the remark this answers is not in the document]
Comment (ckr123) [reply domains/transfer-agreements/TransferAgreements.API/API.IntegrationTests/Controllers/TransferAgreementHistoryEntriesControllerTests.cs:37]: On a deeper look i dont see why you need to make a request to get the wallet deposit endpoint. 
I believe you can use `Some.Base64EncodedWalletDepositEndpoint`

File: domains/transfer-agreements/TransferAgreements.API/API.IntegrationTests/Controllers/TransferAgreementsControllerTest.cs [hunk 12]
```diff
@@ -32,6 +36,21 @@ public class TransferAgreementsControllerTests : IClassFixture<TransferAgreement
         var sub = Guid.NewGuid().ToString();
         factory.WalletUrl = wallet.WalletUrl;
         authenticatedClient = factory.CreateAuthenticatedClient(sub);
+
+        InitializeContext();
+    }
+
+    private void InitializeContext()
```
Comment (ckr123) [inline domains/transfer-agreements/TransferAgreements.API/API.IntegrationTests/Controllers/TransferAgreementsControllerTest.cs:43]: The initialize and dispose. Why are these needed?

File: domains/transfer-agreements/TransferAgreements.API/API.IntegrationTests/Testcontainers/WalletContainer.cs [hunk 13]
```diff
@@ -31,7 +27,7 @@ public WalletContainer()
                 .WithPortBinding(80, true)
                 .WithCommand("--serve", "--migrate")
                 .WithEnvironment("ConnectionStrings__Database", postgresConnectionString)
-                // - This should be implemented correctly, right now it apparantly doesn't use it
+                // - This should be implemented correctly, right now it apparently doesn't use it
```
Comment (ckr123) [inline domains/transfer-agreements/TransferAgreements.API/API.IntegrationTests/Testcontainers/WalletContainer.cs:30]: Why is this here?
If this is something that isnt being used it should be deleted.
If there is some implementation needed we should do it as a TODO with a link to a zenhub issue

File: domains/transfer-agreements/TransferAgreements.API/API.UnitTests/Services/WalletDepositEndpointServiceTests.cs [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+    private Mock<IOptions<ProjectOriginOptions>> mockProjectOriginOptions;
+    private Mock<ILogger<WalletDepositEndpointService>> mockLogger;
+
+    public WalletDepositEndpointServiceTests()
+    {
+        mockProjectOriginOptions = new Mock<IOptions<ProjectOriginOptions>>();
+        mockLogger = new Mock<ILogger<WalletDepositEndpointService>>();
+    }
+
+    private WalletDepositEndpointService CreateService()
```
Comment (ckr123) [inline domains/transfer-agreements/TransferAgreements.API/API.UnitTests/Services/WalletDepositEndpointServiceTests.cs:24]: This should be instantiated in the constructor and use a field variable.

File: domains/transfer-agreements/TransferAgreements.API/API.IntegrationTests/Controllers/TransferAgreementsControllerTest.cs [hunk 15]
```diff
@@ -531,4 +568,36 @@ public async Task CreateWalletDepositEndpoint_ShouldReturnUnauthorized_WhenUnaut
...
+            .PostAsync("api/transfer-agreements/wallet-deposit-endpoint", null);
+
+        var resultData = await result.Content.ReadFromJsonAsync<Dictionary<string, string>>();
+        var base64String = resultData!["result"];
+
+        var transferAgreement = new CreateTransferAgreement(
+            new DateTimeOffset(2123, 3, 3, 3, 3, 3, TimeSpan.Zero).ToUnixTimeSeconds(),
+            new DateTimeOffset(2124, 4, 4, 4, 4, 4, TimeSpan.Zero).ToUnixTimeSeconds(),
+            "12345456",
+            base64String
```
Comment (ckr123) [inline domains/transfer-agreements/TransferAgreements.API/API.IntegrationTests/Controllers/TransferAgreementsControllerTest.cs:588]: Here you can use the `Some.Base64` again

File: domains/transfer-agreements/TransferAgreements.API/API.IntegrationTests/Controllers/TransferAgreementsControllerTest.cs [hunk 16]
```diff
@@ -531,4 +568,36 @@ public async Task CreateWalletDepositEndpoint_ShouldReturnUnauthorized_WhenUnaut
...
+        );
+
+        var response = await authenticatedClient.PostAsJsonAsync("api/transfer-agreements", transferAgreement);
+        response.EnsureSuccessStatusCode();
+
+        var createdTransferAgreement = await response.Content.ReadFromJsonAsync<TransferAgreementDto>();
+
+        await Task.Delay(2000);
```
Comment (ckr123) [inline domains/transfer-agreements/TransferAgreements.API/API.IntegrationTests/Controllers/TransferAgreementsControllerTest.cs:596]: Why a delay here?

File: domains/transfer-agreements/TransferAgreements.API/API.IntegrationTests/Controllers/TransferAgreementsControllerTest.cs [hunk 17]
```diff
@@ -18,17 +20,37 @@
 namespace API.IntegrationTests.Controllers;
 
 [UsesVerify]
-public class TransferAgreementsControllerTests : IClassFixture<TransferAgreementsApiWebApplicationFactory>
+public class TransferAgreementsControllerTests : IClassFixture<TransferAgreementsApiWebApplicationFactory>, IClassFixture<WalletContainer>, IDisposable
 {
     private readonly TransferAgreementsApiWebApplicationFactory factory;
     private readonly HttpClient authenticatedClient;
```
Comment (ckr123) [inline domains/transfer-agreements/TransferAgreements.API/API.IntegrationTests/Controllers/TransferAgreementsControllerTest.cs:26]: I get a lot of warnings on this field.
All the local variables uses the same name i hink either should this be renamed or all the others clients be renamed so there is no confusion on which client you are using.

File: domains/transfer-agreements/TransferAgreements.API/API/Controllers/TransferAgreementsController.cs [hunk 18]
```diff
@@ -20,16 +27,18 @@ public class TransferAgreementsController : ControllerBase
     private readonly ITransferAgreementRepository transferAgreementRepository;
     private readonly IValidator<CreateTransferAgreement> createTransferAgreementValidator;
     private readonly IWalletDepositEndpointService walletDepositEndpointService;
-
+    private readonly IOptions<ProjectOriginOptions> projectOriginOptions;
```
Comment (ckr123) [inline domains/transfer-agreements/TransferAgreements.API/API/Controllers/TransferAgreementsController.cs:30]: Looks like this is not being used

File: domains/transfer-agreements/TransferAgreements.API/API/Controllers/TransferAgreementsController.cs [hunk 19]
```diff
@@ -20,16 +27,18 @@ public class TransferAgreementsController : ControllerBase
...
 
     public TransferAgreementsController(
         ITransferAgreementRepository transferAgreementRepository,
         IValidator<CreateTransferAgreement> createTransferAgreementValidator,
-        IWalletDepositEndpointService walletDepositEndpointService)
+        IWalletDepositEndpointService walletDepositEndpointService,
```
Comment (ckr123) [inline domains/transfer-agreements/TransferAgreements.API/API/Controllers/TransferAgreementsController.cs:35]: I think there is some odd with this now that a service layer is introduced.
We have this controller that has a dependency to the `Data` layer and a `Service` layer now.
Have you considered making a transferagreement service that will handle everything regarding transfer agreements data layer and the wallet deposit endpoint?

File: domains/transfer-agreements/TransferAgreements.API/API.IntegrationTests/Controllers/TransferAgreementsControllerTest.cs (hunk 16 above)
Comment (sahma19) [reply domains/transfer-agreements/TransferAgreements.API/API.IntegrationTests/Controllers/TransferAgreementsControllerTest.cs:596]: Because the builder is too slow build and host the ports for the test containers, so we are giving it a 2 second delay, so it has time to build the container, before it starts searching for its host port.

File: domains/transfer-agreements/TransferAgreements.API/API.IntegrationTests/Testcontainers/WalletContainer.cs (hunk 13 above)
Comment (sahma19) [reply domains/transfer-agreements/TransferAgreements.API/API.IntegrationTests/Testcontainers/WalletContainer.cs:30]: There is no Zenhub Issue. Should i create one then?

File: domains/transfer-agreements/TransferAgreements.API/API/Controllers/TransferAgreementsController.cs (hunk 19 above)
Comment (sahma19) [reply domains/transfer-agreements/TransferAgreements.API/API/Controllers/TransferAgreementsController.cs:35]: No

File: domains/transfer-agreements/TransferAgreements.API/API.IntegrationTests/Controllers/TransferAgreementsControllerTest.cs [hunk 20]
```diff
@@ -531,4 +568,36 @@ public async Task CreateWalletDepositEndpoint_ShouldReturnUnauthorized_WhenUnaut
...
+        );
+
+        var response = await authenticatedClient.PostAsJsonAsync("api/transfer-agreements", transferAgreement);
+        response.EnsureSuccessStatusCode();
+
+        var createdTransferAgreement = await response.Content.ReadFromJsonAsync<TransferAgreementDto>();
+
+        await Task.Delay(2000);
+
+        var createdAgreementInDb = await context.TransferAgreements.FindAsync(createdTransferAgreement.Id);
```
[the remark this answers is not in the document]
Comment (sahma19) [reply domains/transfer-agreements/TransferAgreements.API/API.IntegrationTests/Controllers/TransferAgreementsControllerTest.cs:598]: It can't, since it only returns the transfer agreement in dto format. the dto does not include the ReferenceGuid. That variable is stored in the database, behind the API layer. We need access to the database to access the specific field and its value

File: domains/transfer-agreements/TransferAgreements.API/API/Controllers/TransferAgreementsController.cs (hunk 19 above)
Comment (sahma19) [reply domains/transfer-agreements/TransferAgreements.API/API/Controllers/TransferAgreementsController.cs]: I will create a zenhub issue with a tech-debt label, for refacturing the transfer agreement domain.

File: domains/transfer-agreements/TransferAgreements.API/API.IntegrationTests/Controllers/TransferAgreementsControllerTest.cs [hunk 21]
```diff
@@ -308,13 +320,14 @@ public async Task GetBySubjectId_ShouldReturnTransferAgreements_WhenUserHasTrans
...
                     SenderName = "Producent A/S",
                     SenderTin = senderTin,
-                    ReceiverTin = "87654321"
+                    ReceiverTin = "87654321",
+                    ReceiverReference = Guid.NewGuid()
                 }
             });
 
-        var authenticatedClient = factory.CreateAuthenticatedClient(sub, tin: receiverTin);
+        var client = factory.CreateAuthenticatedClient(sub, tin: receiverTin);
```
Comment (ckr123) [inline domains/transfer-agreements/TransferAgreements.API/API.IntegrationTests/Controllers/TransferAgreementsControllerTest.cs:328]: I've noticed that in almost all test you are creating a "senderClient" with a random sender guid.
Why not create a field variable so you don't have to instantiate a new client on every test when there is no need to specific values.

File: domains/transfer-agreements/TransferAgreements.API/API.UnitTests/Controllers/TransferAgreementsControllerTests.cs [hunk 22]
```diff
@@ -79,7 +80,7 @@ private TransferAgreementsController CreateControllerWithMockedUser()
     [Fact]
     public async Task Create_ShouldCallRepositoryOnce()
     {
-        var request = new CreateTransferAgreement(DateTimeOffset.UtcNow.AddDays(1).ToUnixTimeSeconds(), DateTimeOffset.UtcNow.AddDays(2).ToUnixTimeSeconds(), "13371337");
+        var request = new CreateTransferAgreement(DateTimeOffset.UtcNow.AddDays(1).ToUnixTimeSeconds(), DateTimeOffset.UtcNow.AddDays(2).ToUnixTimeSeconds(), "13371337", "eyJFbmRwb2ludCI6IlNvbWVFbmRwb2ludCIsIlB1YmxpY0tleSI6IkFBQUFBQUFBQUFBQUFBPT0iLCJWZXJzaW9uIjoxfQ==");
```
Comment (ckr123) [inline domains/transfer-agreements/TransferAgreements.API/API.UnitTests/Controllers/TransferAgreementsControllerTests.cs:83]: Why not use `Some.` here?

Comment (sahma19) [reply domains/transfer-agreements/TransferAgreements.API/API.UnitTests/Controllers/TransferAgreementsControllerTests.cs:83]: Changed

File: domains/transfer-agreements/TransferAgreements.API/API.IntegrationTests/Controllers/TransferAgreementsControllerTest.cs (hunk 21 above)
Comment (sahma19) [reply domains/transfer-agreements/TransferAgreements.API/API.IntegrationTests/Controllers/TransferAgreementsControllerTest.cs:328]: I have done what you asked, and only created new clients where relevant.

File: domains/transfer-agreements/TransferAgreements.API/API/Converters/Base64Converter.cs [hunk 23]
```diff
@@ -0,0 +1,61 @@
...
+public class TempWalletDepositEndpoint
+{
+    public string Endpoint { get; set; }
+    public byte[] PublicKey { get; set; }
+    public int Version { get; set; }
+}
+
+public static class Base64Converter
+{
+    public static string ConvertObjectToBase64(object obj)
```
Comment (ckr123) [inline domains/transfer-agreements/TransferAgreements.API/API/Converters/Base64Converter.cs:18]: Should be private

File: domains/transfer-agreements/TransferAgreements.API/API/Converters/Base64Converter.cs [hunk 24]
```diff
@@ -0,0 +1,61 @@
...
+    {
+        var jsonString = JsonSerializer.Serialize(obj);
+        var bytes = Encoding.UTF8.GetBytes(jsonString);
+        var base64String = Convert.ToBase64String(bytes);
+
+        return base64String;
+    }
+
+    public static T? ConvertToObject<T>(string base64String)
```
Comment (ckr123) [inline domains/transfer-agreements/TransferAgreements.API/API/Converters/Base64Converter.cs:27]: Should be private

File: domains/transfer-agreements/TransferAgreements.API/API/Controllers/TransferAgreementsController.cs [hunk 25]
```diff
@@ -56,14 +56,23 @@ public async Task<ActionResult> Create([FromBody] CreateTransferAgreement reques
             SenderId = Guid.Parse(subject),
             SenderName = subjectName,
             SenderTin = subjectTin,
-            ReceiverTin = request.ReceiverTin
+            ReceiverTin = request.ReceiverTin,
```
Comment (ckr123) [inline domains/transfer-agreements/TransferAgreements.API/API/Controllers/TransferAgreementsController.cs:59]: The `,` should be deleted

File: domains/transfer-agreements/TransferAgreements.API/API/Controllers/TransferAgreementsController.cs [hunk 26]
```diff
@@ -56,14 +56,23 @@ public async Task<ActionResult> Create([FromBody] CreateTransferAgreement reques
...
         };
 
         if (await transferAgreementRepository.HasDateOverlap(transferAgreement))
         {
             return Conflict();
         }
 
+        var jwtToken = new JwtToken(User.FindIssuerClaim(), User.FindAudienceClaim(), subject, subjectName);
+        var bearerToken = jwtToken.GenerateToken();
```
[the remark this answers is not in the document]
Comment (ckr123) [reply domains/transfer-agreements/TransferAgreements.API/API/Controllers/TransferAgreementsController.cs:68]: As suggested at review i think it could be implemented here

File: domains/transfer-agreements/TransferAgreements.API/API/Services/WalletDepositEndpointService.cs [hunk 27]
```diff
@@ -0,0 +1,81 @@
...
+        var walletRequest = new CreateReceiverDepositEndpointRequest
+        {
+            Reference = receiverTin,
+            WalletDepositEndpoint = wde
+        };
+
+        var response = await walletServiceClient.CreateReceiverDepositEndpointAsync(walletRequest, headers);
+        return response.ReceiverId;
+    }
+    public Guid ConvertUuidToGuid(ProjectOrigin.Common.V1.Uuid receiverId)
```
[the remark this answers is not in the document]
Comment (sahma19) [reply domains/transfer-agreements/TransferAgreements.API/API/Services/WalletDepositEndpointService.cs]: I removed the separate method and moved its logic to the CreateReceiverDepositEndpoint method

File: domains/transfer-agreements/TransferAgreements.API/API/Services/WalletDepositEndpointService.cs [hunk 28]
```diff
@@ -0,0 +1,81 @@
...
+    }
+
+    public async Task<string> CreateWalletDepositWithToken(JwtToken token)
+    {
+        var bearerToken = token.GenerateToken();
+        var walletDepositEndpoint = await GetWalletDepositEndpoint(bearerToken);
+        return Base64Converter.ConvertWalletDepositEndpointToBase64(walletDepositEndpoint);
+    }
+
+    public async Task<WalletDepositEndpoint> GetWalletDepositEndpoint(string bearerToken)
```
[the remark this answers is not in the document]
Comment (sahma19) [reply domains/transfer-agreements/TransferAgreements.API/API/Services/WalletDepositEndpointService.cs:31]: Changed.

File: domains/transfer-agreements/TransferAgreements.API/API/Services/WalletDepositEndpointService.cs [hunk 29]
```diff
@@ -0,0 +1,81 @@
...
+    private readonly IOptions<ProjectOriginOptions> projectOriginOptions;
+    private readonly ILogger<WalletDepositEndpointService> logger;
+
+    public WalletDepositEndpointService(IOptions<ProjectOriginOptions> options, ILogger<WalletDepositEndpointService> logger)
+    {
+        projectOriginOptions = options;
+        this.logger = logger;
+    }
+
+    public async Task<string> CreateWalletDepositWithToken(JwtToken token)
```
[the remark this answers is not in the document]
Comment (sahma19) [reply domains/transfer-agreements/TransferAgreements.API/API/Services/WalletDepositEndpointService.cs:24]: Renamed.

File: domains/transfer-agreements/TransferAgreements.API/API/Controllers/TransferAgreementsController.cs (hunk 26 above)
[the remark this answers is not in the document]
Comment (sahma19) [reply domains/transfer-agreements/TransferAgreements.API/API/Controllers/TransferAgreementsController.cs:72]: I have made a separate Pull Request for this, and it will be handled there.

File: domains/transfer-agreements/TransferAgreements.API/API/Converters/Base64Converter.cs (hunk 24 above)
Comment (sahma19) [reply domains/transfer-agreements/TransferAgreements.API/API/Converters/Base64Converter.cs:27]: Done

File: domains/transfer-agreements/TransferAgreements.API/API/Controllers/TransferAgreementsController.cs (hunk 25 above)
Comment (sahma19) [reply domains/transfer-agreements/TransferAgreements.API/API/Controllers/TransferAgreementsController.cs]: removed

**Review by ckr123 — APPROVED** [review]
